docs/decisions/blue-green-deployments.md

```diff
@@ -0,0 +1,190 @@
+* it may be desirable to avoid parallel deployments in non-production environments / locally when resources are constrained and a period of down time is acceptable
+* it should be easy to switch other services to reference the new version once testing / initialization has been completed
+
+## Considered Options
+
+There are two aspects to this decision:
+
+* How to deploy the alternative service versions
+* Switching between the two deployments
+
+### Deployment Mechanism
+
+* Deploy different versions on separate hosts
+* Deploy the two versions as separate docker compose projects
+* Manage the different versions within a single docker compose project
+
+### Switching Between Deployments
+
+* Use a shared reverse proxy
+* Use a reverse proxy on each project
+* Target version is configured independently for each referencing service
+* Re-deploy the target service with the normal ports re-mapped
+
+## Decision Outcome
+
+Chosen option is to "Manage the different versions within a single docker compose project" and "Re-deploy the target service with the normal ports re-mapped" as this will work equally well for local and production deployments and does not require other services to be re-configured during the transition.
+
+The short period of downtime during the switch from one version to the next would be no different than for normal deployments.
+
+### Consequences
+
+The chosen option is not strictly a "blue-green" approach but more of a "current-next" option with versioned storage. An example of the steps needed to make a transition is provided below:
+
+An existing `example` service with one volume that maps its internal port 80 to 8080 on the host, e.g.:
+
+```
+services:
+  example:
+    image: ghcr.io/openfoodfacts/example:${TAG}
+    build: .
+    ports:
+      - 8080:80
+    volumes:
+      - dbdata:/var/data
+```
+
+The first step in deploying a new version would be to add a new service and volume and update the old service to use the last build image, e.g.
+
+```
+services:
+  example:
+    image: ghcr.io/openfoodfacts/example:sha_of_last_build_of_old_version
+    ports:
+      - 8080:80
+    volumes:
+      - dbdata:/var/data
+
+  example_next:
+    image: ghcr.io/openfoodfacts/example:${TAG}
+    build: .
+    ports:
+      - 8081:80
+    volumes:
+      - dbdata_v2:/var/data
+```
+This will create a new container with the new image version exposed on a different port with a new volume. The old version will be re-created from the previous build image.
+
+When it is time to switch to the new version the docker compose file would be updated as follows:
+
+```
+services:
+  example:
+    image: ghcr.io/openfoodfacts/example:${TAG}
+    build: .
+    ports:
+      - 8080:80
+    volumes:
+      - dbdata_v2:/var/data
+```
+Note that the dbdata_v2 volume name is retained. Running `docker compose up --remove-orphans` in this case would remove the `example_next` container and replace the old `example` container with the new version.
```

Comment:
In production we use external volumes for volumes that contains data (this is important to be able to do `docker compose down -v` without loosing actual data, but also because loosing automatically created volumes is far too easy).

So you must add:
* modification of external volume name in Makefile
* creation of the new external volume
and after migration:
* removal of old external volume.